test-validator/src/lib.rs

```diff
@@ -482,15 +485,21 @@ impl TestValidatorGenesis {
         let program_path = solana_program_test::find_file(&format!("{program_name}.so"))
             .unwrap_or_else(|| panic!("Unable to locate program {program_name}"));
 
-        self.programs.push(ProgramInfo {
+        self.upgradeable_programs.push(UpgradeableProgramInfo {
             program_id,
-            loader: solana_sdk::bpf_loader::id(),
+            loader: solana_sdk::bpf_loader_upgradeable::id(),
+            upgrade_authority: Pubkey::default(),
             program_path,
         });
         self
     }
 
     /// Add a list of programs to the test environment.
+    #[deprecated(
+        since = "1.16.0",
+        note = "Please use `add_upgradeable_programs_with_path()` instead"
+    )]
+    #[allow(deprecated)]
```

Comment:
Is the intention that this is the last way to use the non-upgradeable loader? It would be easier to just remove `programs` straight up and convert `ProgramInfo` into `UpgradeableProgramInfo`, but I prefer your approach here for more flexibility.

Reply:
Well, yes, I suppose. There is an argument to be made that users that already have programs deployed on a public cluster with BpfLoader2 may still want to run tests with them.
But mostly I just didn't want to delete a method that is part of the public interface. If we're okay with breaking TestValidator and sure that no one will want to load immutable programs anymore, we could strip out this and `ProgramInfo`. Maybe we can do that later, after we see how the upgradeable-default change goes...? 😅